### PR TITLE
OOG error Circuit for `EXP`

### PR DESCRIPTION
**Pull request by silathdiir**
Spec Markdown PR https://github.com/scroll-tech/zkevm-specs/pull/58

### Summary

~~1. Add and export `byte_size_value` function from `ByteSizeGadget` (which is also used to calculate gas cost in `ErrorOOGExpGadget`). And rename original `byte_size` to `byte_size_expr`.~~

2. Implement `ErrorOOGExpGadget` to handle OOG error of `EXP` opcode.

### CODE REVIEW

**Review by roynalnaruto — APPROVED** [review]
LGTM

File: zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let [base, exponent] = [0, 1].map(|idx| block.rws[step.rw_indices[idx]].stack_value());
+
+        let gas_cost = gas_cost_value(exponent);
```
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs:149]: I think we can use` step.gas_cost ` directly to save one method.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            offset,
+            Value::known(F::from(step.gas_left)),
+            Value::known(F::from(gas_cost)),
+        )?;
+        self.restore_context
+            .assign(region, offset, block, call, step, 4)
+    }
+}
+
+fn gas_cost_value(exponent: U256) -> u64 {
```
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs:179]: ditto can remove it.

File: zkevm-circuits/src/evm_circuit/execution/exp.rs [hunk 3]
```diff
@@ -184,7 +184,7 @@ impl<F: Field> ExecutionGadget<F> for ExponentiationGadget<F> {
         );
 
         // Finally we build an expression for the dynamic gas cost.
-        let dynamic_gas_cost = 50.expr() * exponent_byte_size.byte_size();
+        let dynamic_gas_cost = 50.expr() * exponent_byte_size.byte_size_expr();
```
Comment (DreamWuGit) [inline zkevm-circuits/src/evm_circuit/execution/exp.rs:187]: replace 50 with `OpcodeId::EXP.constant_gas_cost().0` ?

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/exp.rs:187]: LGTM with some comments.

File: zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs (hunk 1 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs]: Replaced with `step.gas_cost` in [c4f7176](https://github.com/scroll-tech/zkevm-circuits/pull/323/commits/c4f7176226f2ffe8ae135d0c5c57162005cc9750).

File: zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs (hunk 2 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/error_oog_exp.rs]: Deleted this function in [c4f7176](https://github.com/scroll-tech/zkevm-circuits/pull/323/commits/c4f7176226f2ffe8ae135d0c5c57162005cc9750).

File: zkevm-circuits/src/evm_circuit/execution/exp.rs (hunk 3 above)
Comment (silathdiir) [reply zkevm-circuits/src/evm_circuit/execution/exp.rs]: `OpcodeId::EXP.constant_gas_cost().0 = 10`, it is `static_gas`.

Added a new constant `GasCost::EXP_BYTE_TIMES = 50` (and replaced in both ExpGadget and OogExpGadget).
And add the below comments for gas cost calculation.
```
gas_cost = static_gas + dynamic_gas
static_gas = 10
dynamic_gas = 50 * exponent_byte_size
```

**Comment by silathdiir** [comment]
Thanks for @roynalnaruto and @DreamWuGit 's code review. Moved to upstream https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1232.

**Pull request closed without merging**